enhancement: energy_loss update

dmulash opened the pull request.

RHammond2 commented:
Resolves #10

RHammond2 reviewed (commented):
This is an excellent start @dmulash, thanks for updating all the impacted methods!

I left a lot of comments, but they're mostly relating to the use of an electrical losses method, rather than an input. I have a couple of questions in there as well that'd be great to get your input on, especially while we're modifying the code. I also made a handful of smaller changes myself just to reduce some of the back and forth.

RHammond2 commented on waves/project.py:
```diff
@@ -1690,19 +1808,13 @@ def capacity_factor(
         by : str
             One of "windfarm" or "turbine". Defaults to "windfarm".
         with_losses : bool, optional
```
From the call: delete this argument entirely.

RHammond2 commented on waves/project.py:
```diff
@@ -1446,13 +1438,10 @@ def energy_production(
             units. If None, then the unnormalized energy production is returned, otherwise it must
             be one of "kw", "mw", or "gw". Defaults to None.
         with_losses : bool, optional
-            Use the :py:attr:`loss_ratio` or :py:attr:`Project.loss_ratio` to post-hoc
```
From the call: remove `with_losses` for energy production methods as they should always be included.

dmulash replied:
removed with_losses everywhere, net capacity factor shows the net capacity factor with all the loss types

RHammond2 commented on waves/project.py:
```diff
@@ -1727,7 +1857,6 @@ def capacity_factor(
                 by="turbine",
                 units="kw",
                 with_losses=with_losses,
```
Just to be thorough, also remove this.

RHammond2 commented on waves/project.py:
```diff
@@ -1624,6 +1613,136 @@ def energy_losses(
+
+        return losses.sum(axis=0).sum() / potential.sum(axis=0).sum()
+
+    def technical_loss_ratio(self) -> float:
```
From the call: ignore this suggestion because it doesn't change.

RHammond2 commented on waves/project.py:
```diff
@@ -1624,6 +1613,136 @@ def energy_losses(
+        else:
+            return 0.01  # ORCA -> Hysterisis (1%) for a fixed-bottom project
+
+    def electrical_loss_ratio(self) -> float:
```
From the call: ignore this because it's a constant.

dmulash commented on waves/project.py:
```diff
@@ -271,9 +268,6 @@ class Project(FromDictMixin):
     reinvestment_rate: float = field(
         default=None, validator=attrs.validators.instance_of((float, type(None)))
     )
-    loss_ratio: float = field(
```
Added as optional input in the config file and with default value of 0.0159

dmulash commented on waves/project.py:
```diff
@@ -1624,6 +1613,136 @@ def energy_losses(
+
+        return total_loss_ratio
+
+    def losses_breakdown(self) -> float:
```
Done. project.energy_losses(losses_breakdown = True) gives you the breakdown

dmulash commented:
Hi @RHammond2,

Apart from your suggested changes (eliminate `with_losses` everywhere, substitute `loss_ratio` with `environmental_loss_ratio`, and keep the logic from `energy_potential()` for `wake_losses()`), I have also updated the config files for floating and fixed to include a line specifying the `turbine_type`. Additionally, I modified the example notebook to correct instances where it was calling `with_losses` for breakdowns.

Feel free to adjust as needed!

RHammond2 commented on waves/project.py:
```diff
@@ -1592,19 +1594,25 @@ def energy_losses(
             units="kw",
         )
+
         production = self.energy_production(  # type: ignore
```
I'm realizing now that `energy_production` having losses applied would create a cyclical calculation, so we'll need to reconfigure this calculation just a little bit. So we'd need to start from the the energy potential, then calculate the loss ratios for each category, then apply them to the losses according the ORCA calculation.

Essentially what's happening in the energy production calculation is the application of the wake and availability losses, so we just have to pull them out independentially instead of calling `energy_production`.

RHammond2 replied:
Now, we can just get the energy potential and subtract the combined wake losses, availability, technical losses, and environmental losses (total losses)

dmulash replied:
Adressed by eliminating energy_production and  doing energy_potential * total_loss_ratio

RHammond2 commented on waves/project.py:
```diff
@@ -1592,19 +1594,25 @@ def energy_losses(
+
         losses = potential - production
 
+        # Compute total loss ratio
+        total_loss_ratio = self.total_loss_ratio(environmental_loss_ratio = environmental_loss_ratio)
```
It looks like this is unused, but it should be used to compute `losses = potential * (1 - total_loss_ratio)`.

dmulash replied:
I am guessing you mean losses = potential * total_loss_ratio

RHammond2 commented on waves/project.py:
```diff
@@ -1624,6 +1631,228 @@ def energy_losses(
 
         return losses / self.capacity(per_capacity)
 
+    def wake_losses(
```
I realized there was an error in how this was done, which was mostly my fault, so I fixed that. I would double check the remaining losses calculations, and replicate the process in this method to ensure that both the control flow and the calculations are correct.

RHammond2 commented on waves/project.py:
```diff
@@ -1624,6 +1647,274 @@ def energy_losses(
+
+        return electrical_loss_ratio
+
+    def total_loss_ratio(
```
The docstring should have an extended note (new paragraph betwween the description sentence and parameters section) that outlines the calculation and references ORCA/related paper for the source of the methodology.

RHammond2 commented on waves/project.py:
```diff
@@ -1592,19 +1610,25 @@ def energy_losses(
             by="turbine",
             units="kw",
         )
+
```
This section should be roughly in the form of:

```python
if not losses_breakdown:
    production = ... # the code that's there already
    losses = potentail.copy()
    losses["total_losses"] = potential.values - production.values  # can use a different column name
else:
    # raise an error for if by == "turbine" because detailed breakdowns are only for the wind farm
    environmental_lr = self.environmental_loss_ratio()
    availability_lr = self.availability(which="energy", by="windfarm", frequency="month-year")
    ... # the remaining loss ratios

    base = potential.values
    environmental_losses = base - base * (1 - environmental_lr)
    availability_losses = base - el - base * (1 - environmental_lr) * (1 - availability_lr)
    ... # the remaining losses
    total_losses = ... # full calculation

    losses = potential.copy
    losses = losses.assign(
        environmental_losses=environmental_losses,
        availability_losses=availability_losses,
        ... # the remaining lossses
    ).rename(columns={}) # final step if you think we should have them displayed more nicely

# convert to desired units

# aggregate to apporpriate time scale ("annual", "project")

# apply aep

# apply per_capacity

RHammond2 replied:
From our conversation: only calculate the loss ratio, not the actual losses. (removed from the example above)

RHammond2 commented on waves/project.py:
```diff
@@ -1624,6 +1653,274 @@ def energy_losses(
+
+        return total_loss_ratio
+
+    def losses_breakdown(
```
git pushThis should really be incorporated into the `energy_losses` method using a new keyword argument called `breakdown` because the loss amounts may need to be broken down across time as well.

RHammond2 commented on waves/project.py:
```diff
@@ -1424,14 +1445,12 @@ def energy_production(
         per_capacity: str | None = None,
-        with_losses: bool = False,
-        loss_ratio: float | None = None,
+        environmental_loss_ratio: float | None = None,
```
Basically the section you removed should be replaced with something like:
```python
losses = self.energy_losses(...)  # pass the parameters that were provided to this method in here
production = potential - losses
```

RHammond2 commented on waves/project.py:
```diff
@@ -1472,31 +1508,33 @@ def energy_production(
         if self.floris_results_type == "wind_rose":
-            power = pd.DataFrame(0.0, dtype=float, index=availability.index, columns=["drop"])
-            power = power.merge(
+            energy_gwh = pd.DataFrame(0.0, dtype=float, index=availability.index, columns=["drop"])
```
@RHammond2 fix this to get potential as the base energy, so losses can be subtracted.

dmulash commented:
**Summary of where we left this enhancement:**

The main objective of this enhancement is to be able to output a loss breakdown with the following losses. (please keep in mind that the total losses are calculated as a product of efficiencies as in the following table, so it will not be possible to output the losses in MWh for each category).

Parameter | Value | Units | Note for   Rob
-- | -- | -- | --
Gross   Capacity Factor | 55.48 | % | From FLORIS   = gross_aep / (project_capacity * 8760)
Wake Losses | 5.78 | % | From FLORIS   = (wake_losses/gross_aep)
Availability   Losses | 5.02 | % | From WOMBAT   = (1-availability)
Electrical   Losses | 3.21 | % | From ORCA   (method from aep.py)
Environmental   Losses | 1.67 | % | Input from user   or default value from ORCA
Technical   Losses | 1.00 | % | Default value   from ORCA
Total   System Losses | 15.68 | % | L_total    = 1 - [(1 - L_1) * (1 - L_2) … * (1 - L_n)]
Net   Capacity Factor | 46.78 | % | NCF = GCF *   (1 – L_total)

The [branch](https://github.com/dmulash/WAVES/tree/enhancement/losses) today can output a losses breakdown.
![image](https://github.com/user-attachments/assets/d8dd3204-f368-4252-9cb3-ff3a78febbaa)

The problem is that the net energy production does not match when we use that loss ratio, compared to when we calculate it by turbine and by month. See screenshots below to see that numbers do not match.
![image](https://github.com/user-attachments/assets/cbffff60-a723-474d-b745-0547a93eeef2)
![image](https://github.com/user-attachments/assets/a7d0cd24-742b-4971-b825-f8dc472e702b)

Rob and Daniel explored potential reasons for the mismatch on values and concluded that it might stem from the accumulation of shortened decimal values by month and by turbine. Rob recommended an update to WOMBAT to address this.

dmulash commented:
**This looks good to me, but please note the following caveat:**

The total loss ratio cannot be applied with greater granularity than the "project" level (e.g., "month-year" or "annual") or the "windfarm" level (e.g., "turbine"). This is because the availability loss ratio calculation relies on the base potential energy production from WOMBAT, rather than the energy_potential used in WAVES for the wake loss ratio and net AEP calculations.
